### ISSUE TITLE
make `prefer_collection_literals` less suggestive for sets

### ISSUE THREAD

**Issue by pq** [dart-lang/linter#1425]
The initial implementation flags lots of places in the flutter repo that would be tricky to migrate.  I'd like to review and relax before trying to push the updated lint.

(Note that this gates integration into the SDK.)

**Comment by pq** [comment]
Mulling over a few concrete examples

```dart
final Set<RenderObject> hits = LinkedHashSet<RenderObject>();
```

```dart
final List<String> where = Set<String>.from(skipped).toList()..sort();
```

(and a few related), I'd like to drop the flagging of:

1. `LinkedHashSet` constructors altogether, and
2. `of` and `from` named constructors as well

@bwilkerson:  I think these were the kinds of cases you maybe anticipated in your earlier review?  Any other thoughts?

**Comment by pq** [comment]
If you want to peruse any more of the failures yourself, here's the failing run:

https://ci.chromium.org/p/dart/builders/luci.dart.try/flutter-analyze-try/26

**Comment by bwilkerson** [comment]
> I'd like to drop the flagging of:
> 1. `LinkedHashSet` constructors altogether, and

I'm not sure why. As I understand it, a set literal creates an instance of `LinkedHashSet`, so the two should be consistent.

> `of` and `from` named constructors as well

I think the assist only works for `of` and `from` when the argument is a literal list. We can convert `new Set.from([1, 2, 3])` into `{1, 2, 3}`, but without the spread operator we can't do anything when the argument is _not_ a list literal.

**Comment by pq** [comment]
> I'm not sure why. As I understand it, a set literal creates an instance of `LinkedHashSet`, so the two should be consistent.

Sorry, I should have shared more of my thinking.  It's because of cases like this:

![image](https://user-images.githubusercontent.com/67586/52543486-ee516c00-2d5e-11e9-87c0-8a8000c3e8fe.png)

that provoke an error like:

```dart
error: The set literal type 'Set<String>' isn't of expected type 'HashSet<String>'. The set's type can be changed with an explicit generic type argument or by changing the element types. (strong_mode_invalid_cast_literal_set at ...)
```

For a first cut I think missing cases like this:

```dart
Set<String> s = new HashSet<String>();
```

which I do think should be potentially flagged is OK.

But then maybe it's alright that folks need to address the type error above after applying an assist?

> I think the assist only works for `of` and `from` when the argument is a literal list.

Aha.  Right.  I had this logic as I recall and then talked myself out of needing it. Can't recall now why though... 🤔

**Comment by bwilkerson** [comment]
> But then maybe it's alright that folks need to address the type error above after applying an assist?

I would say no. Our policy on assists has always been that if the code was valid before applying an assist it should be valid after. If the assist can produce invalid code then we should capture the reproduction case (in an issue, or better as a failing test) and fix the bug.
